controllers/singleton_installer.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"sync/atomic"
+)
+
+type InstallStatus int32
+
+const (
+	InstallStatusUnknown InstallStatus = iota
+	InstallStatusInstalling
+	InstallStatusReady
+)
+
+func (i InstallStatus) String() string {
+	switch i {
+	case InstallStatusInstalling:
+		return "Installing"
+	case InstallStatusReady:
+		return "Ready"
+	default:
+		return "Unknown"
+	}
+}
+
+type singletonInstaller struct {
+	status *InstallStatus
+}
+
+type InstallFn func() error
+
+// SingletonInstaller is a simple interface to ensure that only one thread
+// can run an install at a time.
+type SingletonInstaller interface {
+	Install(InstallFn) (InstallStatus, error)
+}
+
+func NewSingletonInstaller() SingletonInstaller {
```

Comment:
Is there any reason we don't just use a mutex here? We seem to be re-implementing it in a way with all this atomic stuff